cpp/src/strings/regex/regcomp.cpp

```diff
@@ -454,41 +457,62 @@ class regex_parser {
           return PLUS_LAZY;
         }
         return PLUS;
-      case '{':  // counted repetition
+      case '{':  // counted repetition: {n,m}
       {
-        if (*exprp < '0' || *exprp > '9') break;
-        const char32_t* exprp_backup = exprp;  // in case '}' is not found
-        char buff[8]                 = {0};
-        for (int i = 0; i < 7 && *exprp != '}' && *exprp != ',' && *exprp != 0; i++, exprp++) {
-          buff[i]     = *exprp;
-          buff[i + 1] = 0;
-        }
-        if (*exprp != '}' && *exprp != ',') {
-          exprp = exprp_backup;
-          break;
-        }
-        sscanf(buff, "%hd", &yy_min_count);
-        if (*exprp != ',')
-          yy_max_count = yy_min_count;
-        else {
-          yy_max_count = -1;
-          exprp++;
-          buff[0] = 0;
-          for (int i = 0; i < 7 && *exprp != '}' && *exprp != 0; i++, exprp++) {
-            buff[i]     = *exprp;
-            buff[i + 1] = 0;
+        if (!std::isdigit(*exprp)) { break; }
+
+        // transform char32 to char until null, delimiter, non-digit or end is reached;
+        // returns the number of chars read/transformed
+        auto transform_until = [](char32_t const* input,
+                                  char32_t const* end,
+                                  char* output,
+                                  std::string_view const delimiters) -> int32_t {
+          int32_t count = 0;
+          while (*input != 0 && input < end) {
+            auto const ch = static_cast<char>(*input++);
+            // if ch not a digit or ch is a delimiter, we are done
+            if (!std::isdigit(ch) || delimiters.find(ch) != delimiters.npos) { break; }
+            *output++ = ch;
+            ++count;
           }
-          if (*exprp != '}') {
-            exprp = exprp_backup;
-            break;
+          *output = 0;  // null-terminate (for the atoi call)
+          return count;
+        };
+
+        auto const exprp_backup  = exprp;  // save in case matching '}' is not found
+        constexpr auto max_read  = 4;      // 3 digits plus the delimiter
+        constexpr auto max_value = 999;    // support only 3 digits
+        std::vector<char> buffer(max_read + 1);
+
+        // get left-side (n) value => min_count
+        exprp += transform_until(exprp, exprp + max_read, buffer.data(), "},");
+        auto count = std::atoi(buffer.data());
+        if ((*exprp != '}' && *exprp != ',') || (count > max_value)) {
```

Comment:
That is kind of a weak link. If the `max_value` changes to a smaller 3-digit value, the check would need to be re-added. This way, this line should never need to change.

Reply:
But would there ever be a reason to use a number that isn't the largest number representable by `max_read` digits? My understanding of the code was that the limitation was solely in place to control the width of the buffer reads.

Reply:
Yes, I'd like to consider limiting `max_value` to something like 255 in the future which would not change `max_read` but require the `count` check.

Reply:
That's fine with me. Just for my edification, what is the benefit of that choice? It doesn't have any performance implications unless someone actually requests a number that large at runtime, right?

Reply:
I'm not sure I understand the question. There must always be a maximum regardless if it is int-max or 999 or whatever. I thought this a reasonable maximum and nicer in the documentation. The number does contribute to the size of the working memory so may affect runtime performance.

Reply:
I am asking whether there is a difference between 255 and 999 if if a user never actually requests a number > 255. Based on 
> The number does contribute to the size of the working memory so may affect runtime performance.
it sounds like the answer is yes? You allocate memory based on the maximum number of repetitions somewhere? In that case, I assume that the amount of memory increases stepwise as the maximum hits crosses powers of two?

Reply:
Nothing that sophisticated. Just maybe store the value in a smaller variable (e.g. uint8).

Reply:
If it doesn’t affect runtime performance, why arbitrarily limit at 999 and not the max value for the size of the repetitions variable? There’s some awkwardness in explaining this arbitrary limit. If it were INT_MAX, I think the docs might not even need to mention it. After all, string columns also have a length limit, right? It might be impossible to reach a repeat limit of INT_MAX.

Reply:
Is just an arbitrary string of numbers entered by a human that is being converted to an integer so some error checking will need to be done since the string of decimal digits could be any length. Is there some reason not to limit it?

Reply:
Discussed offline. I don't think it's worth blocking on this point so I'm fine with accepting a limit of 999. We can change it later if users need more.